crates/configuration/src/configuration.rs

```diff
@@ -1,85 +1,339 @@
 use std::{collections::BTreeMap, path::Path};
 
-use anyhow::ensure;
+use anyhow::{anyhow, ensure};
 use itertools::Itertools;
-use schemars::JsonSchema;
-use serde::Deserialize;
+use mongodb_support::BsonScalarType;
+use ndc_models as ndc;
 
-use crate::{native_procedure::NativeProcedure, read_directory, schema::ObjectType, Schema};
+use crate::{
+    native_procedure::NativeProcedure,
+    native_query::{NativeQuery, NativeQueryRepresentation},
+    read_directory, schema, serialized,
+};
 
-#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
-#[serde(rename_all = "camelCase")]
+#[derive(Clone, Debug, Default)]
 pub struct Configuration {
-    /// Descriptions of collections and types used in the database
-    pub schema: Schema,
+    /// Tracked collections from the configured MongoDB database. This includes real collections as
+    /// well as virtual collections defined by native queries using
+    /// [NativeQueryRepresentation::Collection] representation.
+    pub collections: BTreeMap<String, ndc::CollectionInfo>,
+
+    /// Functions are based on native queries using [NativeQueryRepresentation::Function]
+    /// representation.
+    pub functions: BTreeMap<String, ndc::FunctionInfo>,
+
+    /// In query requests functions and collections are treated as the same, but in schema
+    /// responses they are separate concepts. So we want a set of [CollectionInfo] values for
+    /// functions for query processing, and we want it separate from `collections` for the schema
+    /// response.
+    pub function_collection_infos: BTreeMap<String, ndc::CollectionInfo>,
+
+    /// Procedures are based on native procedures.
+    pub procedures: BTreeMap<String, ndc::ProcedureInfo>,
 
     /// Native procedures allow arbitrary MongoDB commands where types of results are
     /// specified via user configuration.
-    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
     pub native_procedures: BTreeMap<String, NativeProcedure>,
```

Comment:
Likewise, could `procedures` and `native_procedures` be represented as a single map `BTreeMap<String, (ndc::ProcedureInfo, NativeProcedure)>`?

Reply:
Yep that makes sense. But again I think it's a matter of style since there are no cases where we want both values together.